iree/compiler/Conversion/CodegenUtils/ForOpCanonicalization.cpp

```diff
@@ -0,0 +1,155 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+#include "mlir/Dialect/SCF/SCF.h"
+#include "mlir/Dialect/Vector/VectorOps.h"
+#include "mlir/IR/BlockAndValueMapping.h"
+#include "mlir/IR/Builders.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Pass/PassRegistry.h"
+
+// Pass to combine instructions across ForOp boundary. It is common when doing
+// incremental lowering to generate transient ops that cancel each others out.
+// Canonicalization usually clean up those operations. When the value is loop
+// carried, MLIR canonicalization currently doesn't remove the redundant
+// operations.
+//
+// This pass allow to workaround MLIR limitation and does ad hoc clean up of
+// instructions found in IREE. Once we have a more general mechanism in MLIR
+// this pass can be completely removed.
+// This pass does this kind of transformation:
+// ```
+// %21 = vector.shape_cast %20 : vector<4xf32> to vector<1x4xf32>
+// %22 = scf.for %arg3 = %c0 to %c4096 step %c4 iter_args(%arg4 = %21)
+//    -> vector<1x4xf32> {
+//    [...]
+//    %100 = vector.shape_cast %arg4 : vector<1x4xf32> to vector<4xf32>
+//    [...]
+//    %109 = vector.shape_cast %108 : vector<4xf32> to vector<1x4xf32>
+//    scf.yield %109 : vector<1x4xf32>
+//  }
+//  %24 = vector.shape_cast %22 : vector<1x4xf32> to vector<4xf32>
+// ```
+// ->
+// ```
+// %22 = scf.for %arg3 = %c0 to %c4096 step %c4 iter_args(%arg4 = %20)
+//    -> vector<4xf32> {
+//    [...]
+//    scf.yield %108 : vector<4xf32>
+//  }
+// ```
+
+namespace mlir {
+namespace iree_compiler {
+
+namespace {
+class ForOpArgFolding final : public OpRewritePattern<scf::ForOp> {
+ public:
+  using OpRewritePattern<scf::ForOp>::OpRewritePattern;
+
+  Value FoldCarryDep(scf::ForOp forOp, Operation* ivUser,
+                     Operation* ivDef) const {
+    if (auto shapeCast = dyn_cast<vector::ShapeCastOp>(ivUser)) {
+      if (auto souceOp = dyn_cast<vector::ShapeCastOp>(ivDef)) {
+        if (shapeCast.getType() == souceOp.source().getType())
+          return souceOp.source();
+      }
+    } else if (auto extractOp = dyn_cast<vector::ExtractOp>(ivUser)) {
+      if (auto broadcastOp = dyn_cast<vector::BroadcastOp>(ivDef)) {
+        if (extractOp.getType() == broadcastOp.getSourceType())
+          return broadcastOp.source();
+      }
+    }
+    return Value();
+  }
+
+  void transferBody(Block* source, Block* dest, ArrayRef<Value> results,
+                    PatternRewriter& rewriter) const {
+    // Move all operations to the destination block.
+    rewriter.mergeBlocks(source, dest, dest->getArguments());
+    // Replace the yield op by one that returns only the used values.
+    auto yieldOp = cast<scf::YieldOp>(dest->getTerminator());
+    rewriter.updateRootInPlace(
```

Comment:
I am not sure this is an updateRootInPlace. The `PatternRewriter` thinks the `scf.for` is the root. Here you are modifying a yield. You can just modify it in place.